Bazel crash when using `--inject_repository` in 8.5.0

### Description of the bug:


A "duplicate key" error is crashing the Bazel server when running (see repro notes for context);

```sh
'bazel' 'mod' 'dump_repo_mapping' \
  '--tool_tag=bazelbsp:3.2.0' \
  '' \
  '--inject_repository=bazelbsp_aspect=/Users/mele/__repo-root__/.bazelbsp' \
  '--curses=no' \
  '--color=yes' \
  '--noprogress_in_terminal_title'
```

The output is;
```
INFO: Invocation ID: e0d23d2d-c8c8-4f3a-af16-759d67a34085
 no actions running
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'com.google.devtools.build.lib.bazel.bzlmod.BazelDepGraphValue$$Lambda/0x00007f800040cde8@2998aac1' (requested by nodes 'Key{repoName=@@, rootModuleShouldSeeWorkspaceRepos=true}')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:547)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:435)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.IllegalArgumentException: Duplicate key: (__), values: [__, __].
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:483)
        at com.google.common.collect.RegularImmutableTable.checkNoDuplicate(RegularImmutableTable.java:197)
        at com.google.common.collect.SparseImmutableTable.<init>(SparseImmutableTable.java:76)
        at com.google.common.collect.RegularImmutableTable.forOrderedComponents(RegularImmutableTable.java:186)
        at com.google.common.collect.RegularImmutableTable.forCellsInternal(RegularImmutableTable.java:174)
        at com.google.common.collect.RegularImmutableTable.forCells(RegularImmutableTable.java:146)
        at com.google.common.collect.ImmutableTable$Builder.buildOrThrow(ImmutableTable.java:292)
        at com.google.devtools.build.lib.bazel.bzlmod.BazelDepGraphFunction.getExtensionUsagesById(BazelDepGraphFunction.java:136)
        at com.google.devtools.build.lib.bazel.bzlmod.BazelDepGraphFunction.compute(BazelDepGraphFunction.java:75)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:467)
        ... 6 more
```

For convenience, a formatted version of the `IllegalArgumentException` message.
```
Duplicate key: (row=@@//:MODULE.bazel%@bazel_tools//tools/build_defs/repo:local.bzl local_repository, column=<root>),
values: [
    ModuleExtensionUsage{
        extensionBzlFile=//:MODULE.bazel,
        extensionName=@bazel_tools//tools/build_defs/repo:local.bzl local_repository,
        isolationKey=Optional.empty,
        proxies=[
            Proxy{location=<builtin>, proxyName=, containingModuleFilePath=MODULE.bazel, devDependency=true, imports={bazelbsp_aspect=bazelbsp_aspect}}
        ],
        tags=[
            Tag{
                tagName=repo,
                attributeValues=AttributeValues{attributes={
                    "name": "bazelbsp_aspect",
                    "path": "/Users/mele/__repo-root__/.bazelbsp"
                }},
                devDependency=true,
                location=<builtin>
            }
        ],
        repoOverrides={}
    },
    ModuleExtensionUsage{
        extensionBzlFile=//:MODULE.bazel,
        extensionName=@bazel_tools//tools/build_defs/repo:local.bzl local_repository,
        isolationKey=Optional.empty,
        proxies=[
            Proxy{location=/Users/mele/__repo-root__/included_module_dot_bazel_prefix.MODULE.bazel:28:17, proxyName=, containingModuleFilePath=included_module_dot_bazel_prefix.MODULE.bazel, devDependency=false, imports={third_party_repo=third_party_repo}},
            Proxy{location=/Users/mele/__repo-root__/MODULE.bazel:145:17, proxyName=, containingModuleFilePath=MODULE.bazel, devDependency=false, imports={private_repo_one=private_repo_one}},
            Proxy{location=/Users/mele/__repo-root__/MODULE.bazel:150:17, proxyName=, containingModuleFilePath=MODULE.bazel, devDependency=false, imports={private_repo_two=private_repo_two}}],
        tags=[
            Tag{
                tagName=repo,
                attributeValues=AttributeValues{attributes={
                    "path": "third_party_repo_path",
                    "name": "third_party_repo"
                }},
                devDependency=false,
                location=/Users/mele/__repo-root__/included_module_dot_bazel_prefix.MODULE.bazel:28:17
            },
            Tag{
                tagName=repo,
                attributeValues=AttributeValues{attributes={
                    "path": "tools/bazel_rules/config",
                    "name": "private_repo_one"
                }},
                devDependency=false,
                location=/Users/mele/__repo-root__/MODULE.bazel:145:17
            },
            Tag{
                tagName=repo,
                attributeValues=AttributeValues{attributes={
                    "path": "infrastructure/config/starlark/repo_library_testing/test_library",
                    "name": "private_repo_two"
                }},
                devDependency=false,
                location=/Users/mele/__repo-root__/MODULE.bazel:150:17
            }
        ],
        repoOverrides={}
    }
].
```

### Which category does this issue belong to?

External Dependency

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.


This issue was observed with the (new) Bazel plugin for IntelliJ IDEA (v2025.2.8.1). The latest version (v2025.2.9) no longer uses `--inject_repository`, and so is unaffected.

### Which operating system are you running Bazel on?


macOS 26.1

### What is the output of `bazel info release`?


release 8.5.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.


_No response_

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?


```text
NA - private project
```

### If this is a regression, please try to identify the Bazel commit where the bug was introduced with bazelisk --bisect.


Broken in `8.5.0rc4` and `8.5.0`. Since `8.5.0rc3` the only code change is 1f9e46f088225635e8522fd4785766d544751d57 (which I authored 😓).

My guess is eval of `MODULE.bazel` has some additional logic to manage collisions, and now that the order has flipped `--inject_repository` needs it. Although likely a different bit of logic to what tests currently are covering.

I do not know if Bazel 9 is affected as 95dceeaa73c3496dcb6076830c2d17cbee9aafcc has not yet made it's way into a release candidate (and it remains to be seen if our repository is ready enough for v9 to test).

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

--- a/src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileFunction.java
+++ b/src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileFunction.java
@@ -658,8 +658,7 @@ public class ModuleFileFunction implements SkyFunction {
     }
     // Use the innate extension backing use_repo_rule.
     ModuleExtensionUsageBuilder usageBuilder =
-        new ModuleExtensionUsageBuilder(
-            context,
+        context.getOrCreateExtensionUsageBuilder(
             "//:MODULE.bazel",
             "@bazel_tools//tools/build_defs/repo:local.bzl local_repository",
             /* isolate= */ false);
@@ -686,7 +685,6 @@ public class ModuleFileFunction implements SkyFunction {
           "by --inject_repository",
           thread.getCallStack());
     }
-    context.getExtensionUsageBuilders().add(usageBuilder);
   }
 
   /**

--- a/src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java
+++ b/src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java
@@ -468,27 +468,17 @@ public class ModuleFileGlobals {
             .setContainingModuleFilePath(context.getCurrentModuleFilePath());
 
     String extensionBzlFile = normalizeLabelString(context.getModuleBuilder(), rawExtensionBzlFile);
-    var newUsageBuilder =
-        new ModuleExtensionUsageBuilder(context, extensionBzlFile, extensionName, isolate);
 
     if (context.shouldIgnoreDevDeps() && devDependency) {
       // This is a no-op proxy.
-      return new ModuleExtensionProxy(newUsageBuilder, proxyBuilder);
+      return new ModuleExtensionProxy(
+          new ModuleExtensionUsageBuilder(context, extensionBzlFile, extensionName, isolate),
+          proxyBuilder);
     }
 
-    // Find an existing usage builder corresponding to this extension. Isolated usages need to get
-    // their own proxy.
-    if (!isolate) {
-      for (ModuleExtensionUsageBuilder usageBuilder : context.getExtensionUsageBuilders()) {
-        if (usageBuilder.isForExtension(extensionBzlFile, extensionName)) {
-          return new ModuleExtensionProxy(usageBuilder, proxyBuilder);
-        }
-      }
-    }
-
-    // If no such proxy exists, we can just use a new one.
-    context.getExtensionUsageBuilders().add(newUsageBuilder);
-    return new ModuleExtensionProxy(newUsageBuilder, proxyBuilder);
+    return new ModuleExtensionProxy(
+        context.getOrCreateExtensionUsageBuilder(extensionBzlFile, extensionName, isolate),
+        proxyBuilder);
   }
 
   private String normalizeLabelString(InterimModule.Builder module, String rawExtensionBzlFile)
@@ -826,16 +816,9 @@ public class ModuleFileGlobals {
     String extensionName = bzlFile + ' ' + ruleName;
     // Find or create the builder for the singular "innate" extension of this repo rule for this
     // module.
-    for (ModuleExtensionUsageBuilder usageBuilder : context.getExtensionUsageBuilders()) {
-      if (usageBuilder.isForExtension("//:MODULE.bazel", extensionName)) {
-        return new RepoRuleProxy(usageBuilder);
-      }
-    }
-    ModuleExtensionUsageBuilder newUsageBuilder =
-        new ModuleExtensionUsageBuilder(
-            context, "//:MODULE.bazel", extensionName, /* isolate= */ false);
-    context.getExtensionUsageBuilders().add(newUsageBuilder);
-    return new RepoRuleProxy(newUsageBuilder);
+    return new RepoRuleProxy(
+        context.getOrCreateExtensionUsageBuilder(
+            "//:MODULE.bazel", extensionName, /* isolate= */ false));
   }
 
   @StarlarkBuiltin(name = "repo_rule_proxy", documented = false)

--- a/src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java
+++ b/src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java
@@ -151,8 +151,21 @@ public class ModuleThreadContext extends StarlarkThreadContext {
     }
   }
 
-  List<ModuleExtensionUsageBuilder> getExtensionUsageBuilders() {
-    return extensionUsageBuilders;
+  ModuleExtensionUsageBuilder getOrCreateExtensionUsageBuilder(
+      String extensionBzlFile, String extensionName, boolean isolate) {
+    // Isolated extensions have to always get a new builder, non-isolated ones have to reuse an
+    // existing one if it exists so that they all contribute usages to the same row in a table.
+    if (!isolate) {
+      for (var builder : extensionUsageBuilders) {
+        if (builder.isForExtension(extensionBzlFile, extensionName)) {
+          return builder;
+        }
+      }
+    }
+    var newBuilder =
+        new ModuleExtensionUsageBuilder(this, extensionBzlFile, extensionName, isolate);
+    extensionUsageBuilders.add(newBuilder);
+    return newBuilder;
   }
 
   static class ModuleExtensionUsageBuilder {
@@ -189,7 +202,7 @@ public class ModuleThreadContext extends StarlarkThreadContext {
       proxyBuilders.add(builder);
     }
 
-    boolean isForExtension(String extensionBzlFile, String extensionName) {
+    private boolean isForExtension(String extensionBzlFile, String extensionName) {
       return this.extensionBzlFile.equals(extensionBzlFile)
           && this.extensionName.equals(extensionName)
           && !this.isolate;

--- a/src/test/py/bazel/bzlmod/bazel_overrides_test.py
+++ b/src/test/py/bazel/bzlmod/bazel_overrides_test.py
@@ -800,7 +800,7 @@ class BazelOverridesTest(test_base.TestBase):
     )
 
     # --inject_repository _must not_ affect `use_repo_rule` generated repo names
-    #.
+    # .
     _, stdout, _ = self.RunBazel([
         'mod',
         'dump_repo_mapping',
@@ -815,6 +815,28 @@ class BazelOverridesTest(test_base.TestBase):
         '"+local_repository2+injected_repo"',
         '\n'.join(stdout),
     )
+
+  def testInjectRepositoryAndLocalRepository(self):
+    # Regression test for https://github.com/bazelbuild/bazel/issues/27953
+    self.ScratchFile(
+        'MODULE.bazel',
+        [
+            (
+                'local_repository ='
+                ' use_repo_rule("@bazel_tools//tools/build_defs/repo:local.bzl",'
+                ' "local_repository")'
+            ),
+            'local_repository(name = "local_repo", path = "local_repo")',
+        ],
+    )
+    self.ScratchFile('local_repo/REPO.bazel')
+    self.ScratchFile('injected_repo/REPO.bazel')
+
+    self.RunBazel([
+        'mod',
+        'deps',
+        '--inject_repository=injected_repo=%workspace%/injected_repo',
+    ])
 
   def testOverrideRepositoryOnNonExistentRepo(self):
     self.ScratchFile('other_repo/REPO.bazel')

